proposals/call-tags/Overview.md

```diff
@@ -4,183 +4,254 @@
-That is, so long as a call tag's input types are subtypes of the input types of the function, and the output types of the function are subtypes of the output types of the call tag, then an indirect call using that call tag is completely sound.
-This then supports the surface-level feature of various typed OO languages where a subclass/subinterface can refine the signature of a method to either accept a broader range of inputs or (more often) produce a more precise output.
+In light of these issues with interface tables, [Efficient implementation of Java interfaces: Invokeinterface considered harmless](https://dl.acm.org/doi/10.1145/504282.504291) developed a different implementation strategy, which they called interface-method tables.
+In terms of call tags, the strategy works by first associating a call tag and some non-negative integer smaller than, say, 19 with every method of every interface, spreading methods of the same interface as evenly across the 19 viable numbers as possible.
+In the object descriptor, there are then 19 (immutable) `funcref` fields.
+When creating the object descriptor for the class `Bar`, one assigns the `func_switch` to the `funcref` for field `i` that switches on the call tags for each of the interface methods associated with index `i` that `Bar` implements, calling `Bar`'s corresponding implementation of each such interface method.
+To call an interface method with associated call tag `$baz` and index `k` on a given object, one gets the `k`th `funcref` from the object's descriptor and performs `call_with_tag $baz` on it.
+They found that, except in rare scenarios where a class implements many interface methods associated with the same index, this technique makes interface-method dispatch perform just as efficiently as class-method dispatch.
+Our more recent experiments found that this still holds on modern hardware.
 
-### Abstraction
+If `Bar` extends another class `Baz`, each of `Bar`'s' `funcref`s can forward unhandled call tag's to `Baz`'s corresponding `funcref`.
+This can be used to reduce duplication (so that `Bar` isn't also switching on all the interface methods whose implementations are inherited from `Baz`) and to support separate compilation (so that `Bar` does not need to have complete knowledge of all the interfaces that `Baz` implements).
+
+#### Signature Refinement
+
+In languages like Dart and Kotlin allow a `class Super { Object foo() {...} }` to have a subclass `class Sub extends Super { Bar foo(Object o) {...} }`, where the subclass is allowed to refine the signature of `foo` in the superclass (in this case, taking advantage of covariance of return types).
+This poses a type-checking problem for wasm: for `Sub`'s refined v-table's type to be a subtype of `Super`'s v-table's type, one needs structural subtyping of typed function references, which unfortunately [is undecidable with even modest extensions](https://dl.acm.org/doi/10.1145/143165.143228).
```

Comment:
It would be helpful to name the extensions to give a better idea of how relevant this problem would be to WebAssembly. Also, it's not clear to me why we would need structural subtyping of function references. Why wouldn't nominal subtyping work?

Reply:
Ah, in this sentence, the addition of bounded quantification for expressing either methods without having to cast `this` or for supporting generics without having to regularly cast is undecidable.

As for the second question, nominal subtyping might work as a static approach for this specific case; I haven't had the chance to do a broad review of use cases to confirm this. However, the static approach does not address the follow-up use case involving representation specialization. So these might be complementary features.